lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -349,14 +349,25 @@ static ParseResult parseFExtModuleOp(OpAsmParser &parser,
   return parseFModuleOp(parser, result, /*isExtModule:*/ true);
 }
 
-static LogicalResult verifyFModuleOp(FModuleOp module) {
+static LogicalResult verifyFModuleOp(FModuleOp &module) {
   // The parent op must be a circuit op.
-  auto *parentOp = module.getParentOp();
-  if (!parentOp || !isa<CircuitOp>(parentOp)) {
+  auto parentOp = dyn_cast<CircuitOp>(module.getParentOp());
+  if (!parentOp) {
     module.emitOpError("should be embedded into a firrtl.circuit");
     return failure();
   }
 
+  // Disallow recursive instantiation.
+  for (auto &op : *module.getBodyBlock()) {
+    auto inst = dyn_cast<InstanceOp>(op);
```

Comment:
Instead of doing the scan in the verification of the FModuleOp, would it make sense to do this check in the InstanceOp itself?  The check there would just be "check to see if the module being referenced is the same as the module I'm in".

Reply:
This makes sense.

For the check, I'm still thinking that needs to be a symbol table lookup. Is there a good way to get the closest symbol table or do I need to walk back up from instance -> module -> circuit manually?

Reply:
I think something like this should work (it is constant time): ` dyn_cast<FModuleOp>(instance.getParentOp());`

Reply:
Actually, even better: `instance.getParentOfType<CircuitOp>();` should work.

Reply:
Awesome! This is exactly what I was looking for.